File: backend/lite/detach.py
# Copyright (c) 2013, 2014, 2015 MIT Probabilistic Computing Project.
#
# This file is part of Venture.
#
# Venture is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# Venture is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with Venture.  If not, see <http://www.gnu.org/licenses/>.

from venture.lite.node import isConstantNode
from venture.lite.node import isLookupNode
from venture.lite.node import isApplicationNode
from venture.lite.node import isRequestNode
from venture.lite.node import isOutputNode
from venture.lite.omegadb import OmegaDB
from venture.lite.value import SPRef
from venture.lite.scope import isTagOutputPSP
from venture.lite.sp import VentureSPRecord
from venture.lite.consistency import assertTorus, assertTrace

def detachAndExtract(trace, scaffold, compute_gradient = False):
  assertTrace(trace, scaffold)
  assert len(scaffold.border) == 1
  ans = detachAndExtractAtBorder(trace, scaffold.border[0], scaffold, compute_gradient=compute_gradient)
  assertTorus(scaffold)
  return ans

def detachAndExtractAtBorder(trace, border, scaffold, compute_gradient = False):
  """Returns the weight and an OmegaDB.  The OmegaDB contains
  sufficient information to restore the trace, and, if
  compute_gradient is True, to determine the partial derivative of
  the weight with respect to the value at each node.  The latter is
  computed by one level of reverse-mode AD, with the underlying trace
  serving as tape."""
  weight = 0
  omegaDB = OmegaDB()
  for node in reversed(border):
    if scaffold.isAbsorbing(node):
      weight += detach(trace, node, scaffold, omegaDB, compute_gradient)
    else:
      if node.isObservation: weight += getAndUnconstrain(trace,node)
      weight += extract(trace,node,scaffold,omegaDB, compute_gradient)
  return weight,omegaDB

def getAndUnconstrain(trace,node):
  return unconstrain(trace,trace.getConstrainableNode(node))

def unconstrain(trace,node):
  psp,args,value = trace.pspAt(node),trace.argsAt(node),trace.valueAt(node)
  trace.unregisterConstrainedChoice(node)
  psp.unincorporate(value,args)
  weight = psp.logDensity(value,args)
  psp.incorporate(value,args)
  return weight

def detach(trace, node, scaffold, omegaDB, compute_gradient = False):
  weight = unabsorb(trace, node, omegaDB, compute_gradient)
  weight += extractParents(trace, node, scaffold, omegaDB, compute_gradient)
  return weight

def unabsorb(trace, node, omegaDB, compute_gradient = False):
  # we need to pass groundValue here in case the return value is an SP
  # in which case the node would only contain an SPRef
  psp,args,gvalue = trace.pspAt(node),trace.argsAt(node),trace.groundValueAt(node)
  psp.unincorporate(gvalue,args)
  weight = psp.logDensity(gvalue,args)
  if compute_gradient:
    # Ignore the partial derivative of the value because the value is fixed
    (_, grad) = psp.gradientOfLogDensity(gvalue, args)
    omegaDB.addPartials(args.operandNodes + trace.esrParentsAt(node), grad)
  return weight

def extractParents(trace, node, scaffold, omegaDB, compute_gradient = False):
  weight = 0
  for parent in reversed(trace.esrParentsAt(node)):
    weight += extract(trace, parent, scaffold, omegaDB, compute_gradient)
  for parent in reversed(trace.definiteParentsAt(node)):
    weight += extract(trace, parent, scaffold, omegaDB, compute_gradient)
  return weight

def extractESRParents(trace, node, scaffold, omegaDB, compute_gradient = False):
  weight = 0
  for parent in reversed(trace.esrParentsAt(node)):
    weight += extract(trace, parent, scaffold, omegaDB, compute_gradient)
  return weight

def extract(trace, node, scaffold, omegaDB, compute_gradient = False):
  weight = 0
  weight += maybeExtractStaleAAA(trace, node, scaffold, omegaDB, compute_gradient)

  if scaffold.isResampling(node):
    trace.decRegenCountAt(scaffold,node)
    assert trace.regenCountAt(scaffold,node) >= 0
    if trace.regenCountAt(scaffold,node) == 0:
      if isApplicationNode(node):
        if isRequestNode(node):
          weight += unevalRequests(trace, node, scaffold, omegaDB, compute_gradient)
        weight += unapplyPSP(trace, node, scaffold, omegaDB, compute_gradient)
      else:
        trace.setValueAt(node,None)
        assert isLookupNode(node) or isConstantNode(node)
        assert len(trace.parentsAt(node)) <= 1
        if compute_gradient:
          for p in trace.parentsAt(node):
            omegaDB.addPartial(p, omegaDB.getPartial(node)) # d/dx is 1 for a lookup node
      weight += extractParents(trace, node, scaffold, omegaDB, compute_gradient)

  return weight

def maybeExtractStaleAAA(trace, node, scaffold, omegaDB, compute_gradient = False):
  weight = 0
  value = trace.valueAt(node)
  if isinstance(value,SPRef) and value.makerNode != node and scaffold.isAAA(value.makerNode):
    weight += extract(trace, value.makerNode, scaffold, omegaDB, compute_gradient)
  return weight

def unevalFamily(trace, node, scaffold, omegaDB, compute_gradient = False):
  weight = 0
  if isConstantNode(node): pass
  elif isLookupNode(node):
    assert len(trace.parentsAt(node)) == 1
    if compute_gradient:
      for p in trace.parentsAt(node):
        omegaDB.addPartial(p, omegaDB.getPartial(node)) # d/dx is 1 for a lookup node
    trace.disconnectLookup(node)
    trace.setValueAt(node,None)
    weight += extractParents(trace, node, scaffold, omegaDB, compute_gradient)
  else:
    assert isOutputNode(node)
    weight += unapply(trace, node, scaffold, omegaDB, compute_gradient)
    for operandNode in reversed(node.operandNodes):
      weight += unevalFamily(trace, operandNode, scaffold, omegaDB, compute_gradient)
    weight += unevalFamily(trace, node.operatorNode, scaffold, omegaDB, compute_gradient)
  return weight

def unapply(trace, node, scaffold, omegaDB, compute_gradient = False):
  weight = unapplyPSP(trace, node, scaffold, omegaDB, compute_gradient)
  weight += extractESRParents(trace, node, scaffold, omegaDB, compute_gradient)
  weight += unevalRequests(trace, node.requestNode, scaffold, omegaDB, compute_gradient)
  weight += unapplyPSP(trace, node.requestNode, scaffold, omegaDB, compute_gradient)
  return weight

def teardownMadeSP(trace,node,isAAA):
  spRecord = trace.madeSPRecordAt(node)
  assert isinstance(spRecord,VentureSPRecord)
  assert len(spRecord.spFamilies.families) == 0
  trace.setValueAt(node,spRecord)
  if spRecord.sp.hasAEKernel(): trace.unregisterAEKernel(node)
  if isAAA:
    trace.registerAAAMadeSPAuxAt(node,trace.madeSPAuxAt(node))
  trace.setMadeSPRecordAt(node,None)

def unapplyPSP(trace, node, scaffold, omegaDB, compute_gradient = False):
  psp,args = trace.pspAt(node),trace.argsAt(node)
  if isTagOutputPSP(psp):
    scope,block = [trace.valueAt(n) for n in node.operandNodes[0:2]]
    blockNode = node.operandNodes[2]
    trace.unregisterRandomChoiceInScope(scope,block,blockNode)
  if psp.isRandom(): trace.unregisterRandomChoice(node)
  if isinstance(trace.valueAt(node),SPRef) and trace.valueAt(node).makerNode == node:
    teardownMadeSP(trace,node,scaffold.isAAA(node))

  weight = 0
  psp.unincorporate(trace.valueAt(node),args)
  if scaffold.hasLKernel(node):
    weight += scaffold.getLKernel(node).reverseWeight(trace,trace.valueAt(node),args)
    if compute_gradient:
      # TODO Should this take the whole args?  Should it take the esr parents into account?
      (partial, grad) = scaffold.getLKernel(node).gradientOfReverseWeight(trace, trace.valueAt(node), args)
      omegaDB.addPartial(node, partial)
      omegaDB.addPartials(args.operandNodes, grad)
  omegaDB.extractValue(node,trace.valueAt(node))

#  print "unapplyPSP",trace.valueAt(node)

  trace.setValueAt(node,None)
  if compute_gradient and any([scaffold.isResampling(p) or scaffold.isBrush(p) for p in trace.parentsAt(node)]) and not scaffold.hasLKernel(node):
    # Don't need to compute the simulation gradient if the parents are
    # not in the DRG or brush.
    # Not clear how to deal with LKernels.
    # - DeterministicLKernel is probably ok because the nodes they are applied at
    #   do not have interesting parents
    # - DeterministicMakerAAALKernel is probably ok because its gradientOfReverseWeight
    #   method computes the same thing gradientOfSimulate would properly have computed
    grad = psp.gradientOfSimulate(args, omegaDB.getValue(node), omegaDB.getPartial(node))
    omegaDB.addPartials(args.operandNodes + trace.esrParentsAt(node), grad)

  return weight

def unevalRequests(trace, node, scaffold, omegaDB, compute_gradient = False):
  assert isRequestNode(node)
  weight = 0
  request = trace.valueAt(node)
  # TODO I may have the following AD bug: if a request constructs an
  # expression which contains an embedded constant, the constant node
  # will not propagate the partial wrt that constant; and even if it
  # did, I wouldn't know how to backpropagate that partial through the
  # expression generator to the args.
  if request.lsrs and not omegaDB.hasLatentDB(trace.spAt(node)):
    omegaDB.registerLatentDB(trace.spAt(node),trace.spAt(node).constructLatentDB())

  for lsr in reversed(request.lsrs):
    weight += trace.spAt(node).detachLatents(trace.argsAt(node),lsr,omegaDB.getLatentDB(trace.spAt(node)))
    # TODO add gradient information for detached latents?

  for esr in reversed(request.esrs):
    esrParent = trace.popLastESRParent(node.outputNode)
    if trace.numRequestsAt(esrParent) == 0:
      trace.unregisterFamilyAt(node,esr.id)
      omegaDB.registerSPFamily(trace.spAt(node),esr.id,esrParent)
      weight += unevalFamily(trace, esrParent, scaffold, omegaDB, compute_gradient)

  return weight

File: backend/new_cxx/src/detach.cxx
// Copyright (c) 2014 MIT Probabilistic Computing Project.
//
// This file is part of Venture.
//
// Venture is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// Venture is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with Venture.  If not, see <http://www.gnu.org/licenses/>.

#include "detach.h"
#include "node.h"
#include "concrete_trace.h"
#include "scaffold.h"
#include "lkernel.h"
#include "db.h"
#include "sp.h"
#include "psp.h"
#include "sps/scope.h"
#include <iostream>

using std::cout;
using std::endl;

pair<double,shared_ptr<DB> > detachAndExtract(ConcreteTrace * trace,const vector<Node*> & border,shared_ptr<Scaffold> scaffold)
{
  double weight = 0;
  shared_ptr<DB> db(new DB());
  for (vector<Node*>::const_reverse_iterator borderIter = border.rbegin();
       borderIter != border.rend();
       ++borderIter)
  {
    Node * node = *borderIter;
    if (scaffold->isAbsorbing(node))
    {
      ApplicationNode * appNode = dynamic_cast<ApplicationNode*>(node);
      assert(appNode);
      weight += detach(trace,appNode,scaffold,db);
    }
    else
    {
      if (trace->isObservation(node))
      {
        weight += unconstrain(trace,trace->getConstrainableNode(node));
      }
      weight += extract(trace,node,scaffold,db);
    }
  }
  return make_pair(weight,db);
}


double unconstrain(ConcreteTrace * trace,OutputNode * node)
{
  shared_ptr<PSP> psp = trace->getMadeSP(trace->getOperatorSPMakerNode(node))->getPSP(node);
  shared_ptr<Args> args = trace->getArgs(node);
  VentureValuePtr value = trace->getValue(node);

  trace->unregisterConstrainedChoice(node);
  psp->unincorporate(value,args);
  double weight = psp->logDensity(value,args);
  psp->incorporate(value,args);
  return weight;
}

double detach(ConcreteTrace * trace,ApplicationNode * node,shared_ptr<Scaffold> scaffold,shared_ptr<DB> db)
{
  //cout << "detach(" << node << ")" << endl;

  shared_ptr<PSP> psp = trace->getMadeSP(trace->getOperatorSPMakerNode(node))->getPSP(node);
  shared_ptr<Args> args = trace->getArgs(node);
  VentureValuePtr groundValue = trace->getGroundValue(node);

  if (dynamic_pointer_cast<TagOutputPSP>(psp))
  {
    ScopeID scope = trace->getValue(node->operandNodes[0]);
    BlockID block = trace->getValue(node->operandNodes[1]);
    Node * blockNode = node->operandNodes[2];
    trace->unregisterUnconstrainedChoiceInScope(scope,block,blockNode);
  }

  psp->unincorporate(groundValue,args);
  double weight = psp->logDensity(groundValue,args);
  weight += extractParents(trace,node,scaffold,db);
  return weight;
}


double extractParents(ConcreteTrace * trace,Node * node,shared_ptr<Scaffold> scaffold,shared_ptr<DB> db)
{
  double weight = extractESRParents(trace,node,scaffold,db);
  vector<Node*> definiteParents = node->getDefiniteParents();
  for (vector<Node*>::reverse_iterator defParentIter = definiteParents.rbegin();
       defParentIter != definiteParents.rend();
       ++defParentIter)
  {
    weight += extract(trace,*defParentIter,scaffold,db);
  }
  return weight;
}

double extractESRParents(ConcreteTrace * trace,Node * node,shared_ptr<Scaffold> scaffold,shared_ptr<DB> db)
{
  double weight = 0;
  vector<RootOfFamily> esrParents = trace->getESRParents(node);
  for (vector<RootOfFamily>::reverse_iterator esrParentIter = esrParents.rbegin();
       esrParentIter != esrParents.rend();
       ++esrParentIter)
  {
    weight += extract(trace,esrParentIter->get(),scaffold,db);
  }
  return weight;
}

double extract(ConcreteTrace * trace,Node * node,shared_ptr<Scaffold> scaffold,shared_ptr<DB> db)
{
  //cout << "extractOuter(" << node << ")" << endl;
  double weight = 0;
  VentureValuePtr value = trace->getValue(node);

  shared_ptr<VentureSPRef> spRef = dynamic_pointer_cast<VentureSPRef>(value);
  if (spRef && spRef->makerNode != node && scaffold->isAAA(spRef->makerNode))
  {
    weight += extract(trace,spRef->makerNode,scaffold,db);
  }

  if (scaffold->isResampling(node))
  {
    //cout << "extract(" << node << ") = " << trace->getRegenCount(scaffold,node) << endl;
    trace->decRegenCount(scaffold,node);
    assert(trace->getRegenCount(scaffold,node) >= 0);
    if (trace->getRegenCount(scaffold,node) == 0)
    {
      LookupNode * lookupNode = dynamic_cast<LookupNode*>(node);
      RequestNode * requestNode = dynamic_cast<RequestNode*>(node);
      OutputNode * outputNode = dynamic_cast<OutputNode*>(node);
      if (lookupNode) { trace->clearValue(lookupNode); }
      else if (requestNode)
      {
        weight += unevalRequests(trace,requestNode,scaffold,db);
        weight += unapplyPSP(trace,requestNode,scaffold,db);
      }
      else
      {
        assert(outputNode);
        weight += unapplyPSP(trace,outputNode,scaffold,db);
      }
      weight += extractParents(trace,node,scaffold,db);
    }
  }
  return weight;
}

double unevalFamily(ConcreteTrace * trace,Node * node,shared_ptr<Scaffold> scaffold,shared_ptr<DB> db)
{
  //cout << "unevalFamily(" << node << ")" << endl;
  double weight = 0;

  LookupNode * lookupNode = dynamic_cast<LookupNode*>(node);
  ConstantNode * constantNode = dynamic_cast<ConstantNode*>(node);
  OutputNode * outputNode = dynamic_cast<OutputNode*>(node);

  if (constantNode || (outputNode && outputNode->isFrozen) )
  {
    trace->clearValue(node);
  }
  else if (lookupNode)
  {
    trace->disconnectLookup(lookupNode);
    trace->clearValue(lookupNode);
    weight += extractParents(trace,lookupNode,scaffold,db);
  }
  else
  {
    assert(outputNode);
    weight += unapply(trace,outputNode,scaffold,db);
    for (vector<Node*>::reverse_iterator operandNodeIter = outputNode->operandNodes.rbegin();
         operandNodeIter != outputNode->operandNodes.rend();
         ++operandNodeIter)
    {
      weight += unevalFamily(trace,*operandNodeIter,scaffold,db);
    }
    weight += unevalFamily(trace,outputNode->operatorNode,scaffold,db);
  }
  return weight;
}

double unapply(ConcreteTrace * trace,OutputNode * node,shared_ptr<Scaffold> scaffold,shared_ptr<DB> db)
{
  double weight = unapplyPSP(trace,node,scaffold,db);
  weight += extractESRParents(trace,node,scaffold,db);
  weight += unevalRequests(trace,node->requestNode,scaffold,db);
  weight += unapplyPSP(trace,node->requestNode,scaffold,db);
  return weight;
}

void teardownMadeSP(ConcreteTrace * trace,Node * makerNode,bool isAAA,shared_ptr<DB> db)
{
  shared_ptr<VentureSPRecord> spRecord = trace->getMadeSPRecord(makerNode);
  assert(spRecord);
  trace->setValue(makerNode,spRecord);

  shared_ptr<SP> sp = spRecord->sp;
  if (sp->hasAEKernel()) { trace->unregisterAEKernel(makerNode); }
  db->registerMadeSPAux(makerNode,trace->getMadeSPAux(makerNode));
  if (isAAA)
  {
    OutputNode * outputNode = dynamic_cast<OutputNode*>(makerNode);
    assert(outputNode);
    trace->registerAAAMadeSPAux(outputNode,trace->getMadeSPAux(outputNode));
  }
  trace->destroyMadeSPRecord(makerNode);
}

double unapplyPSP(ConcreteTrace * trace,ApplicationNode * node,shared_ptr<Scaffold> scaffold,shared_ptr<DB> db)
{
  //cout << "unapplyPSP(" << node << ")" << endl;
  shared_ptr<PSP> psp = trace->getMadeSP(trace->getOperatorSPMakerNode(node))->getPSP(node);
  shared_ptr<Args> args = trace->getArgs(node);


  // TODO Tag
  if (dynamic_pointer_cast<TagOutputPSP>(psp))
  {
    ScopeID scope = trace->getValue(node->operandNodes[0]);
    BlockID block = trace->getValue(node->operandNodes[1]);
    Node * blockNode = node->operandNodes[2];
    trace->unregisterUnconstrainedChoiceInScope(scope,block,blockNode);
  }

  if (psp->isRandom()) { trace->unregisterUnconstrainedChoice(node); }
  shared_ptr<VentureSPRef> spRef = dynamic_pointer_cast<VentureSPRef>(trace->getValue(node));
  if (spRef && spRef->makerNode == node) { teardownMadeSP(trace,node,scaffold->isAAA(node),db); }

  VentureValuePtr value = trace->getValue(node);

  double weight = 0;
  psp->unincorporate(value,args);
  if (trace->hasLKernel(scaffold,node)) { weight += trace->getLKernel(scaffold,node)->reverseWeight(trace,value,args); }
  db->registerValue(node,value);

  trace->clearValue(node);

  return weight;
}


double unevalRequests(ConcreteTrace * trace,RequestNode * node,shared_ptr<Scaffold> scaffold,shared_ptr<DB> db)
{
  //cout << "unevalRequests(" << node << ")" << endl;

  double weight = 0;

  const vector<ESR>& esrs = trace->getValue(node)->getESRs();
  const vector<shared_ptr<LSR> >& lsrs = trace->getValue(node)->getLSRs();

  Node * makerNode = trace->getOperatorSPMakerNode(node);
  shared_ptr<SP> sp = trace->getMadeSP(makerNode);
  shared_ptr<SPAux> spAux = trace->getMadeSPAux(makerNode);

  if (!lsrs.empty() && !db->hasLatentDB(makerNode)) { db->registerLatentDB(makerNode,sp->constructLatentDB()); }

  for (vector<shared_ptr<LSR> >::const_reverse_iterator lsrIter = lsrs.rbegin();
       lsrIter != lsrs.rend();
       ++lsrIter)
  {
    weight += sp->detachLatents(spAux,*lsrIter,db->getLatentDB(makerNode));
  }


  for (vector<ESR>::const_reverse_iterator esrIter = esrs.rbegin();
       esrIter != esrs.rend();
       ++esrIter)
  {
    RootOfFamily esrRoot = trace->popLastESRParent(node->outputNode);
    if (trace->getNumRequests(esrRoot) == 0)
    {
      trace->unregisterMadeSPFamily(trace->getOperatorSPMakerNode(node),esrIter->id);
      db->registerSPFamily(trace->getMadeSP(trace->getOperatorSPMakerNode(node)),esrIter->id,esrRoot);
      weight += unevalFamily(trace,esrRoot.get(),scaffold,db);
    }
  }
  return weight;
}
